### ISSUE TITLE
Multiplayer "Out of sync"

### ISSUE THREAD

**Issue by elicoten** [Wargus/wargus#321]
I have been trying to play Wargus in multiplayer mode with two computers - one Linux and one Windows.
I have just installed Wargus 2.4.2 and extracted the data on both computers. 
On Linux I installed version 2.4.2-0~201910180445~ubuntu19.10.1 from the [Launchpad PPA](https://launchpad.net/~stratagus/+archive/ubuntu/ppa)
On Windows I downloaded the release installer from [Github](https://github.com/Wargus/wargus/releases/tag/v2.4.2)

I can start the game using the Linux computer as a host (on the Windows machine I'm experiencing #303), and it works fine until the first action happens e.g. attempting to build a building or attack an enemy unit. But as soon as something happens (asides from units just walking around the map), I get a 'network out of sync' message.

I see something about a metaserver. I can't see how to start a meta server I'm just trying to use the "Create a new game" option on the menu.

I'm connecting via a WiFi network, pings between the two computers is pretty good:

    18 packets transmitted, 18 received, 0% packet loss, time 17025ms
    rtt min/avg/max/mdev = 2.150/4.645/21.983/4.644 ms

Is this a known issue - or is there something I can do to provide more debugging information to help track it down?

Thanks in advance for any clues

**Comment by timfel** [comment]
Apparently @ipochto has the same issue. We had some desync issues in the past but they never reoccurred after fixing one particular thing. It seems this is a new issue, since no one touched that code in years. We will have to investigate. In the meantime, are you running with OpenGL or Software mode?

**Comment by elicoten** [comment]
I haven't changed the OpenGL setting and the default seems to be that it's unticked so I presume that's software rendering mode, in full screen. The Linux one froze when I tried to change the full screen resolution so I changed the video mode settings in preferences.lua to match my native resolution.

Is it worth trying in OpenGL mode?

Also the Linux one has been upgraded several times so is it possible that some files/settings from a previous version have been carried forwards causing the problem?

Thanks!

**Comment by timfel** [comment]
In theory, the launcher should check if the extracted files are of the correct version. But there's no check during multiplayer. If the games do not have the exact same extracted data and the exact same version of Stratagus+Wargus, desyncs are quite likely

**Comment by ipochto** [comment]
I catch sync issue when launch on one PC two instances of wargus with same data (`stratagus -d data.Wargus`). Network ports was different of course.

**Comment by ipochto** [comment]
Issue appears in both cases - with and without OpenGL.

**Comment by ljubomirb** [comment]
I think I'm having the same errors.

- When starting "stable" versions, windows clients cannot connect to other windows server, because of lua incompatibility. I tried for a few hours all that I can think of, even to copy everything once installation is finished.

- When starting "nightly", windows client can connect to windows server, but windows server does not react to client saying "ready", and "start game" never appears.

- Only when starting Linux version using PPA as a server, windows nightly clients can connect and game can be started. Game works until you start building something, and then it's out of sync. 

Windows clients were different machines, and Linux machine was Linux Mint livecd version. I tried on LAN, with or without WIFI. I could try to install linux so it would not be livecd version, but I don't think this is an issue. Also, I do not have, at the moment, non-vm linux machine.

Other than that, I'm not sure if diplomacy works, I think other clients in game report _playername_ sent bad command", with _playername_ being the name of a player. This, however I did not try to investigate further, or if it is in connection with out of sync error or not.
I must say, I tried this whole thing yesterday and I'm impressed of how all works well for this iconic game, when playing single player at least. Good work!

**Comment by elicoten** [comment]
Since @ljubomirb mentioned virtual machines - neither of the computers that I was trying with were virtual machines - they were both real machines running Linux or Windows natively.

Also I used the same BIN/CUE image to extract the data for Wargus so it's unlikely there was any discrepancy between the data files (unless the root cause is within the wartool extractor, or a difference between running it on Windows vs Linux).

I'd also like to echo @ljubomirb's thanks as this is a great and impressive effort despite these little quirks with multiplayer!

**Comment by ljubomirb** [comment]
@elicoten I took the time later on to install livecd version to normal VM linux (Mint 19.3, Mate), but no difference as expected.
After trying all that I can think of, I noticed that during install from war2 original, installer shows some errors that are ignored, at the very beginning of unpacking process. Then I suspected that I might have corrupted original, a 1/4 of a century old media, and tried to find some, possibly in better shape on torrents. I was surprised to find it all, but the error (I can only guess, it goes very fast) was the same. But I did not investigate this any further, since in that case, single player game would not work. And since both computers had the same files, I assumed this could not be the culprit.

I'm somehow at the end of my spare time, so I would not be testing this any further, as my only intention was multiplayer game. I still do however appreciate the time you people invest in this, and single player works great, with some minor problems, if one is into that. Good luck fixing this, and thanks again for your time. Bye

**Comment by ipochto** [comment]
> In theory, the launcher should check if the extracted files are of the correct version. But there's no check during multiplayer. If the games do not have the exact same extracted data and the exact same version of Stratagus+Wargus, desyncs are quite likely

One more thing:
[Floating Point Determinism](https://gafferongames.com/post/floating_point_determinism/)
Did this factor count when engine logic writed?

**Comment by timfel** [comment]
Sorry for the delay, at least one out of sync issue has been solved (as soon as a human player clicked on any building that was in the process of being constructed, the game would desync)

**Comment by benhbell** [comment]
Just trying to replicate this with two wargus instances on the same Ubuntu machine and unable to confirm if the issue is fixed with the latest release because I cannot connect locally. I've set them both to run on a different port at the same  `127.0.0.1` ip address.

![Screenshot from 2020-04-26 14-53-23](https://user-images.githubusercontent.com/966644/80316786-c7e5e800-87cd-11ea-9a85-b3229f99fb2a.png)

**Comment by timfel** [comment]
@benhbell can you try with these commands?
```
 stratagus -p -P 4567 -W -v 640x480 -d  ~/.stratagus/data.Wargus/ -c multiplayer -G client,ip=127.0.0.1
```
and
```
stratagus -p -W -v 640x480 -d  ~/.stratagus/data.Wargus/ -c multiplayer -G server,numplayers=2,map="maps/skirmish/multiplayer/(8)farscape2008.smp.gz",fow=0,aiplayers=6,reveal=1
```

**Comment by benhbell** [comment]
This worked. Two linux hosts were able to have a game. I took note to build thing and attack units. 
![Screenshot from 2020-05-08 15-13-48](https://user-images.githubusercontent.com/966644/81451167-e3e47480-9151-11ea-8708-b3f2e05ed0d7.png)
![Screenshot from 2020-05-08 15-07-26](https://user-images.githubusercontent.com/966644/81451169-e47d0b00-9151-11ea-93fe-b7229e0a9096.png)
![Screenshot from 2020-05-08 15-07-17](https://user-images.githubusercontent.com/966644/81451171-e47d0b00-9151-11ea-9886-0e395206184a.png)

**Comment by benhbell** [comment]
I will try windows next

**Comment by benhbell** [comment]
Unable to reproduce with one Mac osx 10.13 and ubuntu 19.04 lts where linux was the host.

**Comment by benhbell** [comment]
I ran into this issue today on a map with three players. As soon as I build my second naval ship we got a network out of sync error in the chats, and then I (host) received bad messages form another player until it crashed. 

I had launched the game from terminal to see if I'd get any relevant logs. this happened twice.
```
DISCLAIMER:
This software is provided as-is.  The author(s) can not be held liable for any
damage that might arise from the use of this software.
Use it at your own risk.

connect to 192.168.1.170:7775 failed
METACLIENT: Unable to connect to host 192.168.1.170:7775
Metaserver stargame...
...metaserver game started
Can't open file 'maps/World Domination BNE.pud.smp.gz': No such file or directory
Gtk-Message: 14:16:38.607: GtkDialog mapped without a transient parent. This is discouraged.

```

Once I got kicked out I received this popup
-----
Stratagus failed to load game data. If you just launched the game without any arguments, this may indicate a bug with the extraction process. Please report this on https://github.com/Wargus/stratagus/issues/new, and please give details, including: operating system, installation path, username, kind of source CD. A possible solution is to remove the hidden folder /home/ceegee/.stratagus/data.Wargus).

**Comment by timfel** [comment]
You say this happened twice. Do you by chance still have the replay?

**Comment by timfel** [comment]
Also, was it both times when you built the second ship?

**Comment by benhbell** [comment]
It was both times when I made my second elven destroyer, I was the host, but my game crashed so I couldn't do a replay. i am going to try an recreate it locally using just myself. Will get back to you

**Comment by benhbell** [comment]
I haven't been able to replicate it

**Comment by timfel** [comment]
Hmmm... elven destroyer. I'll try to read a bit around the code to possibly
find out

**Comment by benhbell** [comment]
Here's a more likely scenario - I noticed that I was given an update to wargus, even though my version still said `2.4.3` both before and after. Is it possible the person I was on with was on a slightly different version? I had them check their updates and run it and they saw wargus as one of the programs updated. Is it possible we were on a minor version difference?

**Comment by timfel** [comment]
It's possible. However, we try to keep things compatible within a release when we do patches. I would keep this open to see if we see it again.

**Comment by timfel** [comment]
I think i found the issue. I accidentally pushed the new version of Wargus with new pathfinding code out, but only linux users got the updated packages, not windows. So Linux and Windows are currently incompatible for 2.4.3. We'll fix that again with the next release.

**Comment by benhbell** [comment]
Yay! Not sure I did much to help, but happy you found it

**Comment by elicoten** [comment]
@timfel Thanks so much for looking into this.

Just to check - would that issue with the differences in pathfinding algorithm between Windows and Linux also affect v2.4.2, as that was the version I was using at the time when I originally experienced this issue?

(It has just occurred to me as I write this that perhaps in some of the earlier comments you may have mentioned other bugs which were extant in v2.4.2 that have also been fixed along the way? I guess either way, the answer is to wait for v2.4.4 and try again. Thanks again!)
